src/JSystem/J3DGraphBase/J3DGD.cpp

```diff
@@ -7,121 +7,129 @@
 }
 
 /* 802D7644-802D76D4       .text J3DGDSetTexImgPtr__F11_GXTexMapIDPv */
-void J3DGDSetTexImgPtr(_GXTexMapID, void*) {
+void J3DGDSetTexImgPtr(GXTexMapID, void*) {
     /* Nonmatching */
 }
 
 /* 802D76D4-802D77A8       .text J3DGDSetTexImgPtrRaw__F11_GXTexMapIDUl */
-void J3DGDSetTexImgPtrRaw(_GXTexMapID, unsigned long) {
+void J3DGDSetTexImgPtrRaw(GXTexMapID, u32) {
     /* Nonmatching */
 }
 
 /* 802D77A8-802D7840       .text J3DGDSetTexTlut__F11_GXTexMapIDUl10_GXTlutFmt */
-void J3DGDSetTexTlut(_GXTexMapID, unsigned long, _GXTlutFmt) {
+void J3DGDSetTexTlut(GXTexMapID, u32, GXTlutFmt) {
     /* Nonmatching */
 }
 
 /* 802D7840-802D7AF8       .text J3DGDLoadTlut__FPvUl11_GXTlutSize */
-void J3DGDLoadTlut(void*, unsigned long, _GXTlutSize) {
+void J3DGDLoadTlut(void*, u32, GXTlutSize) {
     /* Nonmatching */
 }
 
 /* 802D7AF8-802D7DD8       .text J3DGDSetIndTexMtx__F14_GXIndTexMtxIDPA3_fSc */
-void J3DGDSetIndTexMtx(_GXIndTexMtxID, float(*)[3], signed char) {
+void J3DGDSetIndTexMtx(GXIndTexMtxID, Mtx23, s8) {
     /* Nonmatching */
 }
 
 /* 802D7DD8-802D7ED0       .text J3DGDSetIndTexCoordScale__F16_GXIndTexStageID14_GXIndTexScale14_GXIndTexScale14_GXIndTexScale14_GXIndTexScale */
-void J3DGDSetIndTexCoordScale(_GXIndTexStageID, _GXIndTexScale, _GXIndTexScale, _GXIndTexScale, _GXIndTexScale) {
+void J3DGDSetIndTexCoordScale(GXIndTexStageID, GXIndTexScale, GXIndTexScale, GXIndTexScale, GXIndTexScale) {
     /* Nonmatching */
 }
 
 /* 802D7ED0-802D80D0       .text J3DGDSetIndTexOrder__FUl13_GXTexCoordID11_GXTexMapID13_GXTexCoordID11_GXTexMapID13_GXTexCoordID11_GXTexMapID13_GXTexCoordID11_GXTexMapID */
-void J3DGDSetIndTexOrder(unsigned long, _GXTexCoordID, _GXTexMapID, _GXTexCoordID, _GXTexMapID, _GXTexCoordID, _GXTexMapID, _GXTexCoordID, _GXTexMapID) {
+void J3DGDSetIndTexOrder(u32, GXTexCoordID, GXTexMapID, GXTexCoordID, GXTexMapID, GXTexCoordID, GXTexMapID, GXTexCoordID, GXTexMapID) {
     /* Nonmatching */
 }
 
 /* 802D80D0-802D825C       .text J3DGDSetTevOrder__F13_GXTevStageID13_GXTexCoordID11_GXTexMapID12_GXChannelID13_GXTexCoordID11_GXTexMapID12_GXChannelID */
-void J3DGDSetTevOrder(_GXTevStageID, _GXTexCoordID, _GXTexMapID, _GXChannelID, _GXTexCoordID, _GXTexMapID, _GXChannelID) {
+void J3DGDSetTevOrder(GXTevStageID stage, GXTexCoordID coord1, GXTexMapID map1, GXChannelID color1, GXTexCoordID coord2, GXTexMapID map2, GXChannelID color2) {
     /* Nonmatching */
+    static u8 c2r[] = {0, 1, 0, 1, 0, 1, 7, 5, 6, 0, 0, 0, 0, 0, 0, 7};
+
+    GXTexCoordID r31 = coord1 >= 8 ? GX_TEXCOORD0 : coord1;
+    GXTexCoordID r30 = coord2 >= 8 ? GX_TEXCOORD0 : coord2;
+    GDOverflowCheck(5);
+    bool r27 = map1 != GX_TEXMAP_NULL && (map1 & GX_TEXMAP_DISABLE) == 0;
+    bool r26 = map2 != GX_TEXMAP_NULL && (map2 & GX_TEXMAP_DISABLE) == 0;
+    J3DGDWriteBPCmd((stage / 2 + 0x28) << 24 | r31 << 3 | r27 << 6 | c2r[color2 & 0xf] << 19 | r26 << 18 | r30 << 15 | (map2 & 7) << 0xc | c2r[color1 & 0xf] << 7 | r27 << 6 | r31 << 3 | map1 & 7);
```

Comment:
Good lord

Reply:
That part doesn't even match yet. The massive `or` chains are difficult to parse, even on the twilight princess debug rom.

Reply:
You can probably piece it together from https://github.com/devkitPro/libogc/blob/master/libogc/gx.c#L3965-L4008

Reply:
It's the ordering of the terms that gets me. I try to re-order them, and then the compiler will just suddenly change the order of everything. I don't have the patience to trial and error on that.